src/ucp/stream/stream_recv.c

```diff
@@ -8,132 +8,399 @@
+                            ucp_datatype_t datatype,
+                            ucp_stream_recv_callback_t cb)
+{
+    ucp_request_t *req = ucp_request_get(worker);
+    if (ucs_unlikely(req == NULL)) {
+        return UCS_STATUS_PTR(UCS_ERR_NO_MEMORY);
+    }
+
+    req->flags = UCP_REQUEST_FLAG_CALLBACK;
+    req->recv.stream.cb    = cb;
+    req->recv.state.offset = 0;
+    req->recv.stream.count = 0;
+
+    if (ucs_unlikely(UCP_DT_IS_IOV(datatype))) {
+        req->recv.state.dt.iov.iov_offset    = 0;
+        req->recv.state.dt.iov.iovcnt_offset = 0;
+        req->recv.state.dt.iov.iovcnt        = *count;
+        req->recv.state.dt.iov.memh          = UCT_MEM_HANDLE_NULL;
+    }
+
+    req->recv.buffer   = buffer;
+    req->recv.datatype = datatype;
+    req->recv.length   = ucp_dt_length(datatype, *count, buffer,
+                                       &req->recv.state);
+
+    return req;
+}
+
+UCS_PROFILE_FUNC(ucs_status_ptr_t, ucp_stream_recv_nb,
+                 (ep, buffer, count, datatype, cb, flags),
+                 ucp_ep_h ep, void *buffer, size_t *count,
+                 ucp_datatype_t datatype, ucp_stream_recv_callback_t cb,
+                 unsigned flags)
+{
+    ucp_ep_ext_stream_t *ep_stream;
+    ucp_request_t       *req;
+    ucs_status_ptr_t    rdata;
+    size_t              offset;
+    size_t              length;
+    ssize_t             unpacked;
+
+    if (UCP_DT_IS_GENERIC(datatype)) {
+        ucs_error("ucp_stream_recv_nb doesn't support generic datatype");
+        return UCS_STATUS_PTR(UCS_ERR_NOT_IMPLEMENTED);
     }
 
+    UCP_THREAD_CS_ENTER_CONDITIONAL(&ep->worker->mt_lock);
+
+    req = ucp_stream_request_recv_get(ep->worker, buffer, count, datatype, cb);
+    if (ucs_unlikely(!UCS_PTR_IS_PTR(req))) {
+        goto out;
+    }
+
+    ep_stream = ep->ext.stream;
+
+    /* if there are already posted requests, need to queue this one */
+    if (!ucs_queue_is_empty(&ep_stream->reqs)) {
+        ucs_queue_push(&ep_stream->reqs, &req->recv.queue);
+        goto out;
+    }
+
+    /* OK, lets obtain all arrived data which matches the recv size */
+    while (req->recv.state.offset < req->recv.length) {
+        /* Firstly, check if there is not completely processed rdesc */
+        if (ep_stream->rdesc) {
+            rdata  = ucp_stream_rdesc_data(ep_stream->rdesc);
+            offset = ep_stream->rdesc_offset;
+            length = ep_stream->rdesc_len;
+        } else {
+            /* Then dequeue new ones */
+            rdata  = ucp_stream_recv_data_nb_internal(ep_stream, &length);
+            offset = 0;
+            ucs_assert(!UCS_PTR_IS_ERR(rdata));
+        }
+
+        if (UCS_PTR_IS_PTR(rdata)) {
+            unpacked = ucp_stream_data_unpack(UCS_PTR_BYTE_OFFSET(rdata, offset),
+                                              length - offset, req);
+            if (ucs_unlikely(unpacked < 0)) {
+                ucp_request_complete_stream_recv(req, unpacked);
```

Comment:
Don't you need to return an error status here? Currently you would return req pointer